### PR TITLE
[WIP] SKS Keyserver

### PR DESCRIPTION
**Pull request by abizer**
sks-keyserver is running in a docker container provided by an upstream
maintainer, and nginx on the host reverse proxies 80/443 to 11371 (hkp)
I'll expose 11371 itself on nginx as well as another proxied port, and
then work on configuring nginx to reverse proxy the recon port (11370)
as well. the frontend site is not puppeted and is basically this
https://github.com/mattrude/pgpkeyserver-lite cloned into /var/www/html.

### CODE REVIEW

File: modules/ocf_keyserver/files/sks-keyserver.service [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+Description=OCF SKS Keyserver Instance
+Requires=network-online.target docker.service
+After=docker.service
+
+[Service]
+User=root
+# we proxy hkp from a different port so nginx can claim 11371
+ExecStart=/usr/bin/docker run --rm \
+    -v /var/lib/sks:/var/lib/sks \
+    -p 0.0.0.0:11370:11370/tcp \
```
Comment (abizer) [inline modules/ocf_keyserver/files/sks-keyserver.service]: @cg505 was right, this needs to listen directly. nginx proxying the connection causes sks to reject peering attempts because the connections appear to come from the docker-proxy and don't match the contents of the membership file

**Comment by abizer** [comment]
I opened up 11370 and 11371 on the external firewall so this is nominally 'live', with the exception that we aren't peering with anyone. I've sent a request to join the sks-devel list so we can ask for peers. On the bright side, pgp.ocf.io is much, much faster than pgp.mit.edu.

File: modules/ocf_keyserver/templates/sks-web-nginx.erb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    root /var/www/html;
+
+    rewrite ^/stats /pks/lookup?op=stats;
+    rewrite ^/s/(.*) /pks/lookup?search=$1;
+    rewrite ^/search/(.*) /pks/lookup?search=$1;
+    rewrite ^/g/(.*) /pks/lookup?op=get&search=$1;
+    rewrite ^/get/(.*) /pks/lookup?op=get&search=$1;
+    rewrite ^/d/(.*) /pks/lookup?op=get&options=mr&search=$1;
+    rewrite ^/download/(.*) /pks/lookup?op=get&options=mr&search=$1;
```
Comment (dkess) [inline modules/ocf_keyserver/templates/sks-web-nginx.erb]: Why are these needed? Why not use the URLs provided by the webapp?

File: modules/ocf_keyserver/manifests/init.pp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      opts => {
+        chain  => 'PUPPET-INPUT',
+        proto  => ['tcp'],
+        dport  => 11370,
+        action => 'accept',
+      };
+  }
+
+  ocf::firewall::firewall46 {
+    '101 allow sks-hkp':
```
[the remark this answers is not in the document]
Comment (abizer) [reply modules/ocf_keyserver/manifests/init.pp]: they're two very different protocols (hkp vs recon) which is why I separated them

File: modules/ocf_keyserver/files/sks-keyserver.service [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    -p 127.0.0.1:11372:11371/tcp \
+    --name sks-keyserver \
+    zhusj/sks
+Restart=always
+
+[Install]
+WantedBy=docker.service
```
[the remark this answers is not in the document]
Comment (abizer) [reply modules/ocf_keyserver/files/sks-keyserver.service]: all this is copied directly from the docker-registry.service unit - https://github.com/ocf/puppet/blob/master/modules/ocf_docker/files/docker-registry.service

File: modules/ocf_keyserver/templates/sks-web-nginx.erb (hunk 2 above)
Comment (abizer) [reply modules/ocf_keyserver/templates/sks-web-nginx.erb]: the web application itself requires these URLs

File: modules/ocf_keyserver/manifests/init.pp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    ];
+  }
+
+  class { 'nginx':
+    manage_repo  => false,
+    confd_purge  => true,
+    server_purge => true;
+  }
+
+  Class['ocf::ssl::default'] ~> Class['Nginx::Service']
```
[the remark this answers is not in the document]
Comment (abizer) [reply modules/ocf_keyserver/manifests/init.pp]: it should exist from the `class { 'nginx' }` resource earlier, but this service is near-impossible to clean build from puppet alone anyways because generating the database from a keydump needs to be done by hand. If we can figure out the PV/PVC thing, it may be possible to move this service to Kubernetes, though - we'd just need to make sure that the source IP for recon requests survives all the layers of ingress controllers and whatnot

**Comment by abizer** [comment]
things look good and the peering setup works as well so I'm going to merge this
